tracer/src/Datadog.Trace/ClrProfiler/ServerlessInstrumentation/ServerlessMiniAgent.cs

```diff
@@ -0,0 +1,149 @@
+        if (string.IsNullOrEmpty(serverlessMiniAgentPath))
+        {
+            return;
+        }
+
+        if (!File.Exists(serverlessMiniAgentPath))
+        {
+            Log.Error("Serverless Mini Agent does not exist: {Path}", serverlessMiniAgentPath);
+            return;
+        }
+
+        try
+        {
+            Log.Debug("Trying to spawn the Serverless Mini Agent at path: {Path}", serverlessMiniAgentPath);
+            Process process = new Process();
+            process.StartInfo.FileName = serverlessMiniAgentPath;
+            process.StartInfo.UseShellExecute = false;
+            process.StartInfo.RedirectStandardOutput = true;
+            process.OutputDataReceived += new DataReceivedEventHandler(MiniAgentDataReceivedHandler);
+
+            process.Start();
+            process.BeginOutputReadLine();
+        }
+        catch (Exception ex)
+        {
+            Log.Error(ex, "Error spawning the Serverless Mini Agent.");
+        }
+    }
+
+    // Tries to clean Mini Agent logs and log to the correct level, otherwise just logs the data as-is to Info
+    // Mini Agent logs will be prefixed with "[Datadog Serverless Mini Agent]"
+    private static void MiniAgentDataReceivedHandler(object sender, DataReceivedEventArgs outLine)
+    {
+        var data = outLine.Data;
+        if (string.IsNullOrEmpty(data))
+        {
+            return;
+        }
+
+        var logTuple = ProcessMiniAgentLog(data);
+        string level = logTuple.Item1;
+        string processedLog = logTuple.Item2;
+
+        switch (level)
+        {
+            case "ERROR":
+                Log.Error("[Datadog Serverless Mini Agent] {Data}", processedLog);
+                break;
+            case "WARN":
+                Log.Warning("[Datadog Serverless Mini Agent] {Data}", processedLog);
+                break;
+            case "INFO":
+                Log.Information("[Datadog Serverless Mini Agent] {Data}", processedLog);
+                break;
+            case "DEBUG":
+                Log.Debug("[Datadog Serverless Mini Agent] {Data}", processedLog);
+                break;
+            default:
+                Log.Information("[Datadog Serverless Mini Agent] {Data}", data);
+                break;
+        }
+    }
+
+    // Processes a raw log from the mini agent, returning a Tuple of the log level and the cleaned log data
+    // For example, given this raw log:
+    // [2023-06-06T01:31:30Z DEBUG datadog_trace_mini_agent::mini_agent] Random log
+    // This function will return:
+    // ("DEBUG", "Random log")
+    internal static Tuple<string, string> ProcessMiniAgentLog(string rawLog)
+    {
+        int logPrefixCutoff = rawLog.IndexOf("]");
+        if (logPrefixCutoff < 0 || logPrefixCutoff == rawLog.Length - 1)
+        {
+            return Tuple.Create("INFO", rawLog);
+        }
+
+        string level = rawLog.Substring(0, logPrefixCutoff).Split(' ')[1];
```

Comment:
Again not critical if this is not on the hot path, but there are multiple heap allocations here that could be avoided:
- a new string from `Substring()` (here and below)
- a new `string[]` and multiple strings from `Split()`
- potentially a new string from `Trim()` below

The `Split()` in particular caught my attention because it allocates multiple object only to use one of the strings in `array[1]`. We could instead find the position of the first and second `' '` and use that to get a single substring.

We could also use `#if NETCOREAPP` to take advantage of newer .NET features like `Span<char>`. You can see some examples in `W3CTraceContextPropagator` where we avoid `Substring()` in .NET Core: https://github.com/DataDog/dd-trace-dotnet/blob/0ded97871ebb8988fc165012adf6fac93f6149a1/tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs#L251-L301

Reply:
thanks!